github: code coverage workflows

RaduBerinde opened the pull request:
This change adds two GitHub Action workflows which run on each PR. One generates unit test code coverage data, and one publishes that data to a GCS bucket from where Reviewable can access it.

We generate coverage data using `bazel coverage`, but we restrict it to only test the packages that have been modified by the PR.

Two workflows are required for security (the first workflow runs potentially malicious code from a fork); for more details, see https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108485)
<!-- Reviewable:end -->

RaduBerinde commented:
Friendly ping.

rickystewart commented on .github/workflows/code-cover-gen.yml:
```diff
@@ -0,0 +1,77 @@
+          NUM_COMMITS=$(gh pr view $PR --json commits --jq '.commits | length')
+          # This limit must not exceed the checkout fetch-depth above.
+          if [ ${NUM_COMMITS} -ge 10 ]; then
+            echo "PR #$PR has too many commits (${NUM_COMMITS})"
```
Not sure this new requirement is good. We have imposed no such requirement in the past so this is technically a regression. Is the idea just that we detect it if the PR has a billion commits or something?

rickystewart commented on .github/workflows/code-cover-gen.yml:
```diff
@@ -0,0 +1,77 @@
+          CHANGED_PKGS=$(build/ghactions/changed-go-pkgs.sh ${BASE_SHA} ${HEAD_SHA})
+          if [[ $(echo "${CHANGED_PKGS}" | wc -w) -gt 20 ]]; then
+            echo "PR #$PR changes too many packages; skipping code coverage"
+            exit 1
```
If we're going to do this, then you should `exit 0` here so merging your PR is not blocked. Sometimes you have to change 21+ packages at once.

rickystewart commented on build/ghactions/pr-codecov-run-tests.sh:
```diff
@@ -0,0 +1,54 @@
+
+echo "Running tests"
+
+bazel coverage --compilation_mode=fastbuild \
```
`fastbuild` is the default, no reason to specify it. Note that this also results in *stripped* output. Is that undesirable for coverage? Maybe it doesn't matter, I'm not sure. If it does matter you can set `--strip=never`.

rickystewart commented on build/ghactions/pr-codecov-run-tests.sh:
```diff
@@ -0,0 +1,54 @@
+bazel coverage --compilation_mode=fastbuild \
+  --@io_bazel_rules_go//go/config:cover_format=lcov --combined_report=lcov \
+  --instrumentation_filter="//pkg/..." --test_verbose_timeout_warnings \
+  ${targets}
```
I assume you've tested this locally.

rickystewart commented on build/ghactions/pr-codecov-run-tests.sh:
```diff
@@ -0,0 +1,54 @@
+
+bazel coverage --compilation_mode=fastbuild \
+  --@io_bazel_rules_go//go/config:cover_format=lcov --combined_report=lcov \
+  --instrumentation_filter="//pkg/..." --test_verbose_timeout_warnings \
```
`--test_verbose_timeout_warnings` seems irrelevant.

rickystewart commented on build/ghactions/pr-codecov-run-tests.sh:
```diff
@@ -0,0 +1,54 @@
+fi
+
+
+# Find the targets. We need to convert from, e.g.
```
Nit: in `changed-go-pkgs`, we go to efforts to output the package name with the leading `./` (`echo -n "./$path "`), but here we struggle to convert if to something else. If you just output the names without the leading `./` in the first place then there's less back-and-forth conversion to do.

rickystewart reviewed (changes requested):
Mostly this looks good.

We just use the default runner for this right? It's a 2 core machine with like 4GB of RAM or something. My suspicion is it will either simply be insufficient for doing the code coverage analysis, or if it is it'll be extremely slow and swapping all the time.

Let me discuss with the team about opting into bigger runners.

RaduBerinde reviewed (commented):
It is a 2 core with 7GB RAM. I think it will be enough for most PRs given that we're only testing a small set of packages.

Note that this build does NOT block merging of PRs do a failure just mean no coverage data.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108485#-:-NbaeNkH1c-BNg-s9UqN:bwumfir)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @rickystewart, and @srosenberg)

___
*[`.github/workflows/code-cover-gen.yml` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1291787151:-NbaejvLC6X83oi7VxJ8:b-gnr74o) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c7edac6470f3e94e3c2c86c9dcaee84eeb28d67/.github/workflows/code-cover-gen.yml#L41)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

Not sure this new requirement is good. We have imposed no such requirement in the past so this is technically a regression. Is the idea just that we detect it if the PR has a billion commits or something?
</blockquote></details>

This won't block merging. It's because we do a shallow checkout and only fetch a few commits of history. Fetching everything is a good factor slower.
___
*[`.github/workflows/code-cover-gen.yml` line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1291797516:-NbaexM70r5hdQtal8dn:bll5i8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c7edac6470f3e94e3c2c86c9dcaee84eeb28d67/.github/workflows/code-cover-gen.yml#L48)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

If we're going to do this, then you should `exit 0` here so merging your PR is not blocked. Sometimes you have to change 21+ packages at once.
</blockquote></details>

It won't block merging.
___
*[`build/ghactions/pr-codecov-run-tests.sh` line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1291799523:-Nbaf4nA9L29SSsHxGdJ:bmju2ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c7edac6470f3e94e3c2c86c9dcaee84eeb28d67/build/ghactions/pr-codecov-run-tests.sh#L45)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

I assume you've tested this locally.
</blockquote></details>

I tested in some draft PRs


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

A sample build of the first workflow when some go files are changed can be seen in #108245 (the second workflow will only run once merged).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108485#-:-NbakzN11i3WwzxwGjJB:b-xbiki6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @rickystewart, and @srosenberg)

___
*[`build/ghactions/pr-codecov-run-tests.sh` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1291800157:-NbakBwn4ZxlWYgnZ0hu:b-xu7gys) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c7edac6470f3e94e3c2c86c9dcaee84eeb28d67/build/ghactions/pr-codecov-run-tests.sh#L15)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

Nit: in `changed-go-pkgs`, we go to efforts to output the package name with the leading `./` (`echo -n "./$path "`), but here we struggle to convert if to something else. If you just output the names without the leading `./` in the first place then there's less back-and-forth conversion to do.
</blockquote></details>

Good point. I modified the script (which I mostly copied from the teamcity scripts). By the way, do you know why it does the `if ls "$path"/*.go` thing, shouldn't that always be the case if we pass `"pkg/**/*.go` ?
___
*[`build/ghactions/pr-codecov-run-tests.sh` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1291799397:-NbakeMGDK3Pc1GR480Y:b-k2r3mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c7edac6470f3e94e3c2c86c9dcaee84eeb28d67/build/ghactions/pr-codecov-run-tests.sh#L42)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

`fastbuild` is the default, no reason to specify it. Note that this also results in *stripped* output. Is that undesirable for coverage? Maybe it doesn't matter, I'm not sure. If it does matter you can set `--strip=never`.
</blockquote></details>

It seems to work without it from my and @srosenberg's testing, but I added a comment.
___
*[`build/ghactions/pr-codecov-run-tests.sh` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1291799916:-NbakmuV-wZhTXNvHBCs:b-njzfze) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c7edac6470f3e94e3c2c86c9dcaee84eeb28d67/build/ghactions/pr-codecov-run-tests.sh#L44)):*
<details><summary><i>Previously, rickystewart (Ricky Stewart) wrote…</i></summary><blockquote>

`--test_verbose_timeout_warnings` seems irrelevant.
</blockquote></details>

Removed, thanks.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`.github/workflows/code-cover-gen.yml` line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1291797516:-NbeRgxxBarWSy4dDMb1:b-a1cwyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c7edac6470f3e94e3c2c86c9dcaee84eeb28d67/.github/workflows/code-cover-gen.yml#L48)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It won't block merging.
</blockquote></details>

The alternative to failing the job when the conditions aren't met would be to skip subsequent steps. It would show up as success but if you click on it you would see that most steps are skipped. Perhaps that would be preferable? And then a failure would only result from a test failing (which would be something worth investigating).


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Updated to no longer fail the build when we have too many commits or packages. We now record the error in the output json files. I talked with Reviewable folks and they will plumb that error to the UI (next to the umbrella icon). I also increased the fetch depth to 15.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108485#-:-NbfJtq70yOqNdie2AKT:b6racoi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @rickystewart, and @srosenberg)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Ok to merge this?

renatolabs commented on .github/workflows/code-cover-publish.yaml:
```diff
@@ -0,0 +1,58 @@
+    runs-on: ubuntu-latest
+    if: >
+      github.event.workflow_run.event == 'pull_request' &&
+      github.event.workflow_run.conclusion == 'success'
```
So this step will run even if the `generate` step was skipped? Won't `matchArtifact` find no matches (or even crash since it's referencing `[0]` directly).

renatolabs commented on build/ghactions/pr-codecov-run-tests.sh:
```diff
@@ -0,0 +1,49 @@
+
+if [ -z "${packages}" ]; then
+  echo "No packages; skipping"
+  touch "${output_json_file}"
```
Why is this needed?

renatolabs reviewed (commented):
Thanks for the PR! Left some questions but overall looks good.

I wish most / all of this logic was in Go instead of bash scripts -- they are hard to test and it's too easy to accidentally break when we need to change them. PR should still be good as-is but it's something to think about if this becomes a source of pain in the future.

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108485#-:-Nc5Hmit5P9Qrr3Y5OQI:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder, @renatolabs, @rickystewart, and @srosenberg)

___
*[`.github/workflows/code-cover-gen.yml` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1297677811:-Nc5F2ynCnCFdf4kU2v-:bszldbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/afbaf85302d8e94aa60fe96b057c473a71af2ddb/.github/workflows/code-cover-gen.yml#L55)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nits:

* Should we define constants used within this script at the top? `MAX_COMMITS`, `MAX_PACKAGES`. Easier to change them if/when we need to.
* Wouldn't `gt` make more sense here? 15 commits should still be allowed.
</blockquote></details>

I switched to constants. It should be more clear now. The number of commits must be strictly smaller than the fetch depth or the "base" commit won't be accessible.
___
*[`.github/workflows/code-cover-gen.yml` line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1297679727:-Nc5FoGsAlBZPDxNoEWN:b45d9b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/afbaf85302d8e94aa60fe96b057c473a71af2ddb/.github/workflows/code-cover-gen.yml#L89)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Don't we want to skip this step based on `env.SKIP` as well?
</blockquote></details>

No, we want to upload the `.json` files with the error (generated in `skip`). Added a comment.
___
*[`.github/workflows/code-cover-publish.yaml` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1297682465:-Nc5GBZT0FfCoLw5e7q-:b-6pcpn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/afbaf85302d8e94aa60fe96b057c473a71af2ddb/.github/workflows/code-cover-publish.yaml#L19)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

So this step will run even if the `generate` step was skipped? Won't `matchArtifact` find no matches (or even crash since it's referencing `[0]` directly).
</blockquote></details>

We always upload artifacts even if we skip generation.
___
*[`build/ghactions/pr-codecov-run-tests.sh` line 10 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108485#gh-1297687634:-Nc5Fwe-3Y2KX4LY9PcE:bb98sh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/afbaf85302d8e94aa60fe96b057c473a71af2ddb/build/ghactions/pr-codecov-run-tests.sh#L10)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Why is this needed?
</blockquote></details>

We want to upload the json files, even if they're empty. Otherwise, we'll see an error in Reviewable.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
github: code coverage workflows

This change adds two GitHub Action workflows which run on each PR. One
generates unit test code coverage data, and one publishes that data to
a GCS bucket from where Reviewable can access it.

We generate coverage data using `bazel coverage`, but we restrict it
to only test the packages that have been modified by the PR.

Two workflows are required for security (the first workflow runs
potentially malicious code from a fork); for more details, see
https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

Epic: none
Release note: None

RaduBerinde commented:
bors r+

RaduBerinde commented:
@rickystewart PTAL - the "changes requested" status is blocking merge.

rickystewart reviewed (approved):
OK with merging for now. We can keep an eye on the workflow and see how performance and everything looks.

RaduBerinde commented:
Thank you!

bors r+
